In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv(
    r"C:\Users\Hp\Desktop\Travel-Tourism-Customer-Retention-Analysis\data\processed\hotel_bookings_final_cleaned.csv"
)

df.shape

(87370, 36)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87370 entries, 0 to 87369
Data columns (total 36 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           87370 non-null  object 
 1   is_canceled                     87370 non-null  int64  
 2   lead_time                       87370 non-null  int64  
 3   arrival_date_year               87370 non-null  int64  
 4   arrival_date_month              87370 non-null  object 
 5   arrival_date_week_number        87370 non-null  int64  
 6   arrival_date_day_of_month       87370 non-null  int64  
 7   stays_in_weekend_nights         87370 non-null  int64  
 8   stays_in_week_nights            87370 non-null  int64  
 9   adults                          87370 non-null  int64  
 10  children                        87370 non-null  float64
 11  babies                          87370 non-null  int64  
 12  meal                            

## Dataset Structure

The final cleaned dataset is examined to verify column names,
data types, and record counts before Exploratory Data Analysis.

In [4]:
df.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces 

In [5]:
df.isnull().sum().sum()

np.int64(0)

In [6]:
df.duplicated().sum()

np.int64(11)

In [16]:
duplicates = df[df.duplicated()]
duplicates.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,total_stay_nights,total_guests,booking_changed,has_previous_cancellation,special_request_flag
5288,Resort Hotel,1,106,2016,July,30,21,1,3,2,...,227.0,0,0,Canceled,2016-06-13,4,4.0,0,0,0
21863,Resort Hotel,0,4,2016,August,32,4,0,3,2,...,227.0,0,0,Check-Out,2016-08-07,3,2.0,0,0,0
21972,Resort Hotel,0,50,2016,August,32,5,2,4,2,...,227.0,0,0,Check-Out,2016-08-11,6,2.0,1,0,0
21974,Resort Hotel,0,50,2016,August,32,5,2,4,2,...,227.0,0,0,Check-Out,2016-08-11,6,2.0,1,0,0
22001,Resort Hotel,0,17,2016,August,33,8,1,3,2,...,227.0,0,2,Check-Out,2016-08-12,4,2.0,0,0,1


In [7]:
engineered_features = [
    'total_stay_nights',
    'total_guests',
    'booking_changed',
    'has_previous_cancellation',
    'special_request_flag'
]

df[engineered_features].head()

,total_stay_nights,total_guests,booking_changed,has_previous_cancellation,special_request_flag
0,0,2.0,1,0,0
1,0,2.0,1,0,0
2,1,1.0,0,0,0
3,1,1.0,0,0,0
4,2,2.0,0,0,1


In [17]:
df = df.drop_duplicates()

df.shape

(87359, 36)

In [8]:
df.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,...,agent,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests,total_stay_nights,total_guests,booking_changed,has_previous_cancellation,special_request_flag
count,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,...,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000,87370.000000
mean,0.274969,79.911686,2016.210290,26.839533,15.815360,1.005448,2.625798,1.875953,0.138675,0.010828,...,81.011194,0.749788,105.355488,0.084251,0.698638,3.631246,2.025455,0.181996,0.019286,0.497814
std,0.446501,86.056471,0.686116,13.673204,8.834842,1.031974,2.053692,0.626473,0.455933,0.113613,...,109.941520,10.017213,49.182928,0.281571,0.831946,2.763579,0.794309,0.385844,0.137528,0.499998
min,0.000000,0.000000,2015.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,-6.380000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,11.000000,2016.000000,16.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,...,9.000000,0.000000,72.000000,0.000000,0.000000,2.000000,2.000000,0.000000,0.000000,0.000000
50%,0.000000,49.000000,2016.000000,27.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,...,9.000000,0.000000,98.100000,0.000000,0.000000,3.000000,2.000000,0.000000,0.000000,0.000000
75%,1.000000,125.000000,2017.000000,37.000000,23.000000,2.000000,4.000000,2.000000,0.000000,0.000000,...,234.000000,0.000000,134.000000,0.000000,1.000000,5.000000,2.000000,0.000000,0.000000,1.000000
max,1.000000,737.000000,2017.000000,53.000000,31.000000,19.000000,50.000000,55.000000,10.000000,10.000000,...,535.000000,391.000000,227.000000,8.000000,5.000000,69.000000,55.000000,1.000000,1.000000,1.000000


In [18]:
df.duplicated().sum()

np.int64(0)

## Numerical Data Validation

Statistical summaries are reviewed to ensure
all numeric variables contain reasonable values.

In [19]:
df.to_csv(
    r"C:\Users\Hp\Desktop\Travel-Tourism-Customer-Retention-Analysis\data\processed\hotel_bookings_validated.csv",
    index=False
)

In [9]:
df['adr'].describe()

count    87370.000000
mean       105.355488
std         49.182928
min         -6.380000
25%         72.000000
50%         98.100000
75%        134.000000
max        227.000000
Name: adr, dtype: float64

In [10]:
print("Maximum ADR:", df['adr'].max())
print("Minimum ADR:", df['adr'].min())

Maximum ADR: 227.0
Minimum ADR: -6.38


In [11]:
df['is_canceled'].value_counts()

is_canceled
0    63346
1    24024
Name: count, dtype: int64

In [12]:
df['is_canceled'].value_counts(normalize=True) * 100

is_canceled
0    72.503148
1    27.496852
Name: proportion, dtype: float64

## Cancellation Distribution

The target variable distribution is examined to understand
the balance between cancelled and non-cancelled bookings.

In [13]:
df.dtypes

hotel                              object
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                 object
assigned_room_type                 object
booking_changes                     int64
deposit_type                       object
agent                             

In [14]:
df.dtypes.value_counts()

int64      20
object     12
float64     4
Name: count, dtype: int64

In [20]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 87359
Columns: 36


In [21]:
df.to_csv(
    r"C:\Users\Hp\Desktop\Travel-Tourism-Customer-Retention-Analysis\data\processed\hotel_bookings_validated.csv",
    index=False
)

In [22]:

import os

os.listdir(
    r"C:\Users\Hp\Desktop\Travel-Tourism-Customer-Retention-Analysis\data\processed"
)

['hotel_bookings_feature_engineered.csv',
 'hotel_bookings_final_cleaned.csv',
 'hotel_bookings_validated.csv']

## Findings

- The final dataset contains 87,359 records and 36 columns.
- No missing values were found.
- 11 duplicate records were identified and removed.
- Engineered features were successfully validated.
- ADR outlier treatment was verified.
- The dataset is ready for Exploratory Data Analysis in Week 2.

## Day 6 Conclusion

The final dataset was successfully validated through data quality checks.

Missing values were checked, duplicate records were removed, engineered features were validated, and ADR treatment results were verified.

The validated dataset is now ready for Week 2 Exploratory Data Analysis and business insight generation.